idaes/generic_models/properties/core/eos/virial.py

```diff
@@ -0,0 +1,797 @@
+[1] J. M Prausnitz, R, N Lichtenthaler, and E. G de Azvedo,
+    Molecular thermodynamics of Fluid-phase Equilibruim 3rd ed. Prentice-Hall, Nj, 1998
+**Any standard chemical engineering thermodynamic textbook
+
+Author: Akula Paul
+
+"""
+
+from copy import deepcopy
+from pyomo.environ import (Expression,
+                           log, exp,
+                           value, NonNegativeReals,
+                           Var, units as pyunits)
+from pyomo.common.config import ConfigBlock, ConfigValue, Bool
+
+from idaes.core.util.exceptions import PropertyNotSupportedError
+from idaes.generic_models.properties.core.generic.utility import (
+    get_method, get_component_object as gcobj)
+from idaes.generic_models.properties.core.eos.eos_base import EoSBase
+from idaes.core.util.misc import set_param_from_config
+from idaes.generic_models.properties.core.pure.RPP4 import entr_mol_ig_comp
+
+
+VirialConfig = ConfigBlock()
+
+# virial eos options
+VirialConfig.declare("use_pseudocritical_rules", ConfigValue(
+    default=True,
+    domain=Bool,
+    description='''Flag indicating whether pseudocritical rules should be used
+    when applicable''',
+    doc='''Flag indicating whether to compute mixture property based on
+    composition and critical properties of pure components:
+**default** - True.
+**Valid values:** {
+**True** - use pseudocritical_rule: mixing rule based on composition and
+critical properties of pure components.
+**False** - use mixing rule based on composition and corresponding
+pure component properties.}'''))
+
+
+class enth_mol_ig_comp():
+    # This fixes the bug in RPP4.Since its just a specific method,recreating it
+    # here
+    @staticmethod
+    def build_parameters(cobj):
+        if not hasattr(cobj, "cp_mol_ig_comp_coeff_A"):
+            cp_mol_ig_comp.build_parameters(cobj)
+
+        units = cobj.parent_block().get_metadata().derived_units
+        cobj.enth_mol_form_vap_comp_ref = Var(
+            doc="Vapor phase molar heat of formation @ Tref",
+            units=units["energy_mole"])
+        if cobj.parent_block().config.include_enthalpy_of_formation:
+            set_param_from_config(cobj, param="enth_mol_form_vap_comp_ref")
+        else:
+            cobj.enth_mol_form_vap_comp_ref.set_value(0.0)
+
+    @staticmethod
+    def return_expression(b, cobj, T):
+        # Specific enthalpy
+        T = pyunits.convert(T, to_units=pyunits.K)
+        Tref = pyunits.convert(b.params.temperature_ref, to_units=pyunits.K)
+
+        units = b.params.get_metadata().derived_units
+
+        h = (pyunits.convert(
+            (cobj.cp_mol_ig_comp_coeff_D / 4) * (T**4 - Tref**4) +
+            (cobj.cp_mol_ig_comp_coeff_C / 3) * (T**3 - Tref**3) +
+            (cobj.cp_mol_ig_comp_coeff_B / 2) * (T**2 - Tref**2) +
+            cobj.cp_mol_ig_comp_coeff_A * (T - Tref), units["energy_mole"]) +
+            cobj.enth_mol_form_vap_comp_ref)
+
+        return h
+
+
+class entr_mol_ig_comp_G_H_ref(entr_mol_ig_comp):
```

Comment:
@andrewlee94  Yes, I agree we need a discussion on this especially for clarity between the standard entropy reference (absolute entropy) that users may want to set. For consistency with chemical reactions and data on heat formation and Gibbs energy of formation, this will be required.  This is critical for the framework.